### PR TITLE
Update Regex to Support All ASDF Versions for the supported distributions in tool-versions File

### PR DESCRIPTION
**Pull request by aparnajyothi-y**
**Description:**
This Pull request improves the regex pattern used in the tool-versions file to ensure compatibility with all ASDF versions across the supported distributions. The goal is to provide better flexibility and robustness when parsing tool version entries, handling variations in version formats more effectively.

**Related issue:**
[#719](https://github.com/actions/setup-java/issues/719)

**Check list:**
- [ ✔️] Mark if documentation changes are required.

### CODE REVIEW

File: dist/cleanup/index.js [hunk 1]
```diff
@@ -77289,6 +77289,14 @@ const { isUint8Array, isArrayBuffer } = __nccwpck_require__(9830)
 const { File: UndiciFile } = __nccwpck_require__(8511)
 const { parseMIMEType, serializeAMimeType } = __nccwpck_require__(685)
 
+let random
+try {
+  const crypto = __nccwpck_require__(6005)
+  random = (max) => crypto.randomInt(0, max)
+} catch {
+  random = (max) => Math.floor(Math.random(max))
```
[the remark this answers is not in the document]
Comment (annettejanewilson) [reply dist/cleanup/index.js:77297]: Never mind, this is autogenerated, isn't it?

[the remark this answers is not in the document]
Comment (annettejanewilson) [reply dist/cleanup/index.js:77297]: For reference, I think this is from updating to undici 5.28.5. It looks like it has been [fixed](https://github.com/nodejs/undici/pull/4023) but only released in 6.21.2 and 7.4.0.

File: src/util.ts [hunk 2]
```diff
@@ -133,7 +133,7 @@ export function getVersionFromFileContent(
   const versionFileName = getFileName(versionFile);
   if (versionFileName == '.tool-versions') {
     javaVersionRegExp =
-      /^(java\s+)(?:\S*-)?v?(?<version>(\d+)(\.\d+)?(\.\d+)?(\+\d+)?(-ea(\.\d+)?)?)$/m;
+      /^(java\s+)(?:\S*-)?v?(?<version>(\d+)(\.\d+)?(\.\d+)?(\+\S+)?(\.\d+)?(-ea(\.\d+)?)?(\.LTS)?)$/m;
```
Comment (annettejanewilson) [inline src/util.ts:136]: Is the `.LTS` functional here? As far as I can tell, for the Java versions that have `.LTS` in them, it always gets consumed by the `(\+\S+)?` part, since it matches all non-spaces after a `+`, leaving the rest of the regular expression vestigial. Perhaps that part wasn't expected to match so much?

I tested this regex against the output of `asdf list all java`. I deduplicated those versions that only differ by their numbers.

## Already matched, still match
```
adoptopenjdk-11.0.15+10
adoptopenjdk-jre-11.0.15+10
graalvm-community-17.0.7
liberica-16+36
liberica-11.0.10+9
liberica-javafx-16+36
liberica-javafx-11.0.10+9
liberica-jre-16+36
liberica-jre-11.0.10+9
liberica-jre-javafx-16+36
liberica-jre-javafx-11.0.10+9
liberica-lite-16+36
liberica-lite-11.0.10+9
microsoft-11.0.15
openjdk-17
openjdk-17.0.1
oracle-17
oracle-17.0.1
oracle-graalvm-21
oracle-graalvm-17.0.7
sapmachine-17
sapmachine-0.0.0
sapmachine-jre-17
sapmachine-jre-0.0.0
semeru-jre-openj9-23+37_openj9-0.47.0
semeru-jre-openj9-11.0.15+10_openj9-0.32.0
semeru-jre-openj9-11.0.16.1+1_openj9-0.33.1
semeru-openj9-23+37_openj9-0.47.0
semeru-openj9-11.0.15+10_openj9-0.32.0
semeru-openj9-11.0.16.1+1_openj9-0.33.1
temurin-11.0.15+10
temurin-jre-11.0.15+10
zulu-11.43.1017
zulu-javafx-11.45.27
zulu-jre-11.45.27
zulu-jre-javafx-11.45.27
```
## Newly matched
```
adoptopenjdk-21.0.0+35.0.LTS
adoptopenjdk-jre-21.0.0+35.0.LTS
corretto-8.322.06.4
graalvm-22.1.0+java11
microsoft-11.0.16.1
openjdk-18.0.1.1
oracle-17.0.3.1
sapmachine-11.0.16.1
sapmachine-jre-11.0.16.1
semeru-jre-openj9-23.0.1+11_openj9-0.49.0-m2
semeru-openj9-23.0.1+11_openj9-0.49.0-m2
temurin-21.0.0+35.0.LTS
temurin-jre-21.0.0+35.0.LTS
zulu-8.52.0.23
zulu-javafx-8.52.0.23
zulu-jre-8.52.0.23
zulu-jre-javafx-8.52.0.23
```
## Still not matched
```
adoptopenjdk-19.0.0-beta+36.0.202208190932
adoptopenjdk-jre-19.0.0-beta+36.0.202208190932
corretto-11.0.15.9.1
jetbrains-17.0.4.1b646.8
jetbrains-11.0.16b2043.64
jetbrains-21b212.1
jetbrains-jre-17.0.4.1b646.8
jetbrains-jre-11.0.16b2043.64
jetbrains-jre-21b212.1
kona-8.0.12.b1
liberica-11.0.14.1+1
liberica-8u282+8
liberica-javafx-11.0.14.1+1
liberica-javafx-8u282+8
liberica-jre-11.0.14.1+1
liberica-jre-8u282+8
liberica-jre-javafx-11.0.14.1+1
liberica-jre-javafx-8u282+8
liberica-lite-11.0.14.1+1
liberica-lite-8u302+8
mandrel-23.1.3.1-Final+java21
microsoft-11.0.14.9.1
sapmachine-18-internal.0
sapmachine-19-snapshot
sapmachine-20-snapshot.35
sapmachine-11.0.19-snapshot.1
sapmachine-17.0.3.0.1
sapmachine-jre-18-internal.0
sapmachine-jre-19-snapshot
sapmachine-jre-20-snapshot.35
sapmachine-jre-11.0.19-snapshot.1
sapmachine-jre-17.0.3.0.1
temurin-19.0.0-beta+36.0.202208190932
temurin-jre-19.0.0-beta+36.0.202208190932
zulu-8.62.0.19_1
zulu-11.66.15_1
zulu-javafx-11.66.15_1
zulu-jre-11.66.15_1
zulu-jre-javafx-11.66.15_1
```
Notable causes of remaining misses are:
* having 5 number segments in the version in the case of corretto, microsoft and sapmachine
* having 4 number segments in the version and a `+` separated build metadata section, in the case of liberica
* having a suffix like `_1` for patch versions in the case of zulu
* having a `-` separated pre-release segment like "beta", "snapshot" or "internal" in the case of adoptopenjdk and sapmachine
* having a `b` or `u` separated build number, which seems to be the case for jetbrains and some versions of liberica

I hope that's helpful.

Comment (annettejanewilson) [reply src/util.ts:136]: Oh, in case you want to repeat the analysis, I used jq. The regex engine is close enough for one like this.
```
asdf list all java |
    sed 's/^/java /' |
    jq -cRn \
        --arg oldregex '^(java\s+)(?:\S*-)?v?(?<version>(\d+)(\.\d+)?(\.\d+)?(\+\d+)?(-ea(\.\d+)?)?)$' \
        --arg newregex '^(java\s+)(?:\S*-)?v?(?<version>(\d+)(\.\d+)?(\.\d+)?(\+\S+)?(\.\d+)?(-ea(\.\d+)?)?(\.LTS)?)$' \
        '[
            inputs|
            {
                version:.,
                matches_new:test($newregex),
                matches_old:test($oldregex)
            }
        ]|
        unique_by(.version|=gsub("[0-9]+";"0"))|
        .[]'
```

**Comment by rnorth** [comment]
@aparnajyothi-y we'd be very excited to see this get released - is there anything we can do to help? (testing?)

It looks like there's currently a check failing but it's apparently hitting a GitHub API rate limit. Presumably unrelated to the actual changes though:

```
Fetching version info for GraalVM EA builds from 'https://api.github.com/repos/graalvm/oracle-graalvm-ea-builds/contents/versions/24-ea.json?ref=main' failed with the error: API rate limit exceeded for 13.105.117.124. (But here's the good news: Authenticated requests get a higher rate limit. Check out the documentation for more details.)
```

**Comment by aparnajyothi-y** [comment]
Hello @rnorth, thank you again for your continued interest and support!

We’ve completed the regex and documentation updates, aiming for broad coverage of the formats in your test suite while clearly documenting known limitations such as jetbrains-21b212.1. As discussed earlier, we’ve prioritized maintainability and clarity over overfitting the regex to rare edge cases.

The recent failures appear to be caused by unauthenticated GitHub API rate limits and are unrelated to the core changes in this PR. That said, we’d greatly appreciate your feedback on whether the updated pattern works well against your test cases.

Please let us know if the current implementation meets your expectations so we can proceed with the review and merging process.

**Comment by rnorth** [comment]
Thanks @aparnajyothi-y

It looks like you've changed the regex a few times over the past few weeks, so I thought I should test it out.

I've tried the updated regex ([source link](https://github.com/actions/setup-java/pull/767/files#diff-3294a832ea2276e554177e0b3007cc2d401c082912c7fbde49fa09141bf1aed1R136)):
```
/^java\s+(?:\S*-)?(?<version>\d+(?:\.\d+){0,2}(?:[+_]\d+)?)(?=(?:[-_](?!\d))|$)(?:[-_.](?:ea|LTS|beta|snapshot|internal|b\d+|\d+[a-z]*))*$/im;
```

I've tried this with @annettejanewilson's [helpful `jq` command](https://github.com/actions/setup-java/pull/767#discussion_r2014806034) and [by hand](https://regex101.com/r/oAJbmH/1) - unfortunately it seems like it's failing to match things like `java temurin-21.0.0+35.0.LTS` (or any other Temurin 21 releases).

Temurin is one of the most mainstream distributions, and 21 the most recent LTS version of it - so I'd say this is not just an edge case, but possibly one of the most important distribution versions to be able to match.

There are some other distribution/versions that fail to match that seem highly likely to catch people out:
* LTS versions of `adoptopenjdk`
* Any version of `corretto`

So I think it's unfortunate but the regex's changes since @annettejanewilson's comments seem to have reduced its coverage. 😬  I don't know the reasons for the changes, but they feel like a regression.

I'm wondering separately: looking at all the `src/distributions/*/installer.ts` files, it looks like after all this regex work is done there's corresponding code that has to convert the versions _back_ into what seems like the original string from the `.tool-versions` file. I haven't got time to try it right now but maybe there's an obvious alternative way to approach this: parse the `.tool-versions` file _just enough_ to get the distribution name, and then give the whole raw version part to the relevant `installer.ts` `*Distribution` to find and install. Maybe this should be considered if there are conflicting constraints stopping regex from working...

**Comment by rnorth** [comment]
Thanks @aparnajyothi-y and I'm sorry for being slow to reply.

Yes, I've tried the new regex again with @annettejanewilson's [jq command](https://github.com/actions/setup-java/pull/767#discussion_r2014806034) and it looks good to me - very broad coverage of the ASDF versions list. Thanks for your work on this!

**Comment by rnorth** [comment]
I'm itching to see this in a released version - is there an ETA?
Thank you 🙇
